### ISSUE TITLE
Model getting replicated on 4 CPU instead of being distributed

### ISSUE THREAD

**Issue by dgaleareading** [horovod/horovod#822]
Hi,

I am new to Horovod and so to get to grips with it, I am trying to adapt a [Cat vs Dogs classification](https://github.com/fchollet/deep-learning-with-python-notebooks/blob/master/5.2-using-convnets-with-small-datasets.ipynb) from the Deep Learning from Python book to run on 4 CPUs on my laptop to speed it up.

My code now is:
```

import keras
from keras import optimizers
from keras import layers
from keras import models
from keras.preprocessing.image import ImageDataGenerator
import matplotlib.pyplot as plt
import os
from keras import backend as K
import math
import tensorflow as tf
import horovod.keras as hvd

hvd.init()

config = tf.ConfigProto()
config.gpu_options.allow_growth = True
config.gpu_options.visible_device_list = str(hvd.local_rank())
K.set_session(tf.Session(config=config))

print(hvd.size(), hvd.rank(), hvd.local_rank())

base_dir = '/home/daniel/Desktop/proper_train'
train_dir = os.path.join(base_dir, 'train')
validation_dir = os.path.join(base_dir, 'validation')

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu',
                        input_shape=(150, 150, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.summary()

opt = keras.optimizers.RMSprop()
opt = hvd.DistributedOptimizer(opt)

model.compile(loss=keras.losses.binary_crossentropy,
              optimizer=opt,
              metrics=['accuracy'])

# All images will be rescaled by 1./255
train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

callbacks = [
    # Horovod: broadcast initial variable states from rank 0 to all other processes.
    # This is necessary to ensure consistent initialization of all workers when
    # training is started with random weights or restored from a checkpoint.
    hvd.callbacks.BroadcastGlobalVariablesCallback(0),
]

train_generator = train_datagen.flow_from_directory(
        # This is the target directory
        train_dir,
        # All images will be resized to 150x150
        target_size=(150, 150),
        batch_size=20,
        # Since we use binary_crossentropy loss, we need binary labels
        class_mode='binary')

validation_generator = test_datagen.flow_from_directory(
        validation_dir,
        target_size=(150, 150),
        batch_size=20,
        class_mode='binary')

history = model.fit_generator(
      train_generator,
      steps_per_epoch=25,
      epochs=30,
      validation_data=validation_generator,
      validation_steps=50)

acc = history.history['acc']
val_acc = history.history['val_acc']
loss = history.history['loss']
val_loss = history.history['val_loss']

epochs = range(len(acc))

plt.plot(epochs, acc, 'bo', label='Training acc')
plt.plot(epochs, val_acc, 'b', label='Validation acc')
plt.title('Training and validation accuracy')
plt.legend()

plt.figure()

plt.plot(epochs, loss, 'bo', label='Training loss')
plt.plot(epochs, val_loss, 'b', label='Validation loss')
plt.title('Training and validation loss')
plt.legend()

plt.show()
```

I noticed that the model was being replicated 4 times instead of training being distributed across the 4 CPUs, so I am printing out the size, rank and local_rank using `print(hvd.size(), hvd.rank(), hvd.local_rank())`. I would expect size to be 4 for all cases, rank to range from 0 to 3 and local rank to be 0 in all cases, to show distributed learning instead of replicated models,  but I am getting `1, 0, 0` in all cases.

My question is: am I understanding how Horovod works wrong or am I doing something work for the model not getting distributed? If the latter, any ideas on how I might fix it?

**Comment by alsrgv** [comment]
@dgaleareading, it's likely that you have two versions of MPI installed in your system, and that you've compiled Horovod with a different version of MPI than the one you're using.  Can you reinstall Horovod, while making sure that proper MPI's `mpirun` and `mpicxx` are in $PATH?

**Comment by dgaleareading** [comment]
I have reinstalled everything as suggested and have now managed to get `4, 0, 0`, `4, 1, 1`, `4, 2, 2`, `4, 3, 3`. Also, when I do `which mpirun`, I get `/usr/local/bin/mpirun` and for `which mpicxx`, I get `/usr/local/bin/mpicxx`. Also, `echo $PATH` gets `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin` so everythin is loaded up as it should be (I think). Despite all of this, I am still getting 4 models running instead of one running on 4 cores.

**Comment by alsrgv** [comment]
@dgaleareading, when you use Horovod, you'll get one copy of the model running per compute device.  DL frameworks like TensorFlow have their own mechanism to use many cores w/o having to resort to distributed training - but you can now take your code and run it across servers.

**Comment by dgaleareading** [comment]
Very well. Would Horovod then train each of the models on a different part of the data? If not, how is the speedup achieved? Also, how would I then finish up with one final model?

**Comment by alsrgv** [comment]
@dgaleareading, it would if you configure your data pipeline in such a way.  Since you're using `train_datagen.flow_from_directory`, dataset will do shuffling for you and if you have multiple workers reading dataset for `M/(K*B)` steps, where `M` is the number of examples, `K` is the number of workers and `B` is batch size, you should have reasonably good coverage of the dataset.  You can then specify [steps_per_epoch](https://github.com/horovod/horovod/blob/master/examples/keras_imagenet_resnet50.py#L166) like we do in our example.

There is a more deterministic approach to data partitioning by using [tf.data.Dataset.shard()](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shard) API.  We have tried both approaches before and found they perform approximately the same, so it's generally ok to use whatever is easier.

**Comment by dgaleareading** [comment]
Great! Finally, would I be correct in saying that the model in each rank has the same weights at the end, i.e. does Horovod do the averaging itself, so that I can evaluate the model from any rank and get the same accuracy?

**Comment by alsrgv** [comment]
@dgaleareading, they should be approximately the same - we do not exchange batch norm statistics across GPUs, so there could be a little bit of variation there.

**Comment by dgaleareading** [comment]
@alsrgv, I have run a few experiments with steps_per_epoch as you suggested and am concerned that the accuracy isn't what I expect.

When i run with 1 CPU, I get an accuracy of ~0.75 on the training set ([1cpu.txt](https://github.com/horovod/horovod/files/2871270/1cpu.txt)); with 5CPUs, I get an accuracy of 0.5325 ([5cpu.txt](https://github.com/horovod/horovod/files/2871269/5cpu.txt)); with 10CPUs, I get an accuracy of 0.6 ([10cpu.txt](https://github.com/horovod/horovod/files/2871268/10cpu.txt)) and with 20CPUs, I get an accuracy of .55 ([20cpu.txt](https://github.com/horovod/horovod/files/2871267/20cpu.txt)). I have also attached the [code](https://github.com/horovod/horovod/files/2871277/toy_model.txt) I used. Surely, the point of Horovod is that you still get the same accuracy when compared to a serial run but in a much shorter run?

Another concern is that all the jobs take around the same amount of time, which defeats the purpose of using distribution in the first place. Am I doing something or understanding something wrong?

**Comment by dgaleareading** [comment]
I have now run your [advanced example](https://github.com/horovod/horovod/blob/master/examples/keras_mnist_advanced.py) without any changes on 1, 5, 10, 15 CPUs and got the following results:
```
15 cores - Run time: 915 sec - Test loss: 0.018879091751002126; Test accuracy: 0.9939
10 cores - Run time: 1024 sec - Test loss: 0.015202352570415906; Test accuracy: 0.9954
5 cores - Run time: 1259 sec - Test loss: 0.01590518005306376; Test accuracy: 0.9946
1 cores - Run time: 949sec - Test loss: 0.018812042461237615; Test accuracy: 0.9941
```
Surely this shouldn't be happening as no speedup is obtained?

I am running this on an LSF HPC cluster from inside a Singularity container using the following jobscript:

```
#BSUB -cwd /home/users/dgalea/junk
#BSUB -n 10
#BSUB -W 24:00 
#BSUB -o %J.lo 
#BSUB -e %J.err  
#BSUB -J tm_20_hv
#BSUB -q par-multi
#BSUB –R “rusage[mem=51200]”

/apps/singularity/bin/singularity shell ~/containers/container.img<<EOF
cd /mounts/code
mpirun    -np 10 -bind-to none -map-by slot     -x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH -x PATH     -mca pml ob1 -mca btl ^openib     python3.6 toy_model2.py
EOF
```

where toy_model2.py is your advanced example.

**Comment by alsrgv** [comment]
@dgaleareading, you generally need a lot of steps to train deep learning model.

Because of that, Horovod generally helps when your training takes days or weeks and allows you to cut it to hours, but it's going to struggle to cut down training that takes few minutes to seconds.

Additionally, TensorFlow does not work very well if you give it just a single core.  I would recommend trying to run TensorFlow with 10-20 cores per process and then scale to multiple processes to achieve a speedup.
